src/libexpr/symbol-table.hh

```diff
@@ -70,15 +71,21 @@ public:
 class SymbolTable
 {
 private:
-    typedef std::unordered_set<string> Symbols;
-    Symbols symbols;
+    std::unordered_map<std::string_view, Symbol> symbols;
+    std::list<string> store;
 
 public:
     Symbol create(std::string_view s)
     {
-        // FIXME: avoid allocation if 's' already exists in the symbol table.
-        std::pair<Symbols::iterator, bool> res = symbols.emplace(std::string(s));
-        return Symbol(&*res.first);
+        // Most symbols are looked up more than once, so we trade off insertion performance
```

Comment:
It appears that in C++20, searching using an `std::string_view` in a set of `std::string`s now works. At least, this compiles:
```
-        // FIXME: avoid allocation if 's' already exists in the symbol table.
-        std::pair<Symbols::iterator, bool> res = symbols.emplace(std::string(s));
+        std::pair<Symbols::iterator, bool> res = symbols.emplace(s);
```

So that avoids storing a redundant `std::string_view`. But it should be checked that this doesn't introduce a lot of implicit conversions in every comparison (https://stackoverflow.com/a/53530846).

Reply:
would've used those features if nix were C++20, but we didn't want to change it from 17 to 20 just for this. if that's a reasonable thing to do here we'll happily do it!